agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1411 @@
+use super::super::INSTANCE_COUNT_METRIC;
+use super::{
+    constants::SHARED_INSTANCE_OFFLINE_GRACE_PERIOD_SECS,
+    device_plugin_builder::{DevicePluginBuilder, DevicePluginBuilderInterface},
+    device_plugin_service,
+    device_plugin_service::{
+        get_device_instance_name, InstanceConnectivityStatus, InstanceInfo, InstanceMap,
+    },
+    embedded_discovery_handlers::get_discovery_handler,
+    registration::{
+        DiscoveryHandlerConnectivityStatus, DiscoveryHandlerDetails, RegisteredDiscoveryHandlerMap,
+        DISCOVERY_HANDLER_OFFLINE_GRACE_PERIOD_SECS, EMBEDDED_DISCOVERY_HANDLER_ENDPOINT,
+    },
+    streaming_extension::StreamingExt,
+};
+use akri_discovery_utils::discovery::v0::{
+    discovery_client::DiscoveryClient, Device, DiscoverRequest, DiscoverResponse,
+};
+use akri_shared::{
+    akri::configuration::KubeAkriConfig,
+    k8s,
+    os::env_var::{ActualEnvVarQuery, EnvVarQuery},
+};
+use blake2::{
+    digest::{Update, VariableOutput},
+    VarBlake2b,
+};
+use log::{error, trace};
+#[cfg(test)]
+use mockall::{automock, predicate::*};
+use std::{collections::HashMap, convert::TryFrom, sync::Arc, time::Instant};
+use tokio::sync::{broadcast, mpsc};
+use tonic::{
+    transport::{Endpoint, Uri},
+    Status,
+};
+
+/// StreamType provides a wrapper around the two different types of streams returned from embedded
+/// or internal discovery handlers and ones running externally.
+pub enum StreamType {
+    Internal(mpsc::Receiver<std::result::Result<DiscoverResponse, Status>>),
+    External(tonic::Streaming<DiscoverResponse>),
+}
+
+/// A DiscoveryOperator is created for each Configuration that is applied to the cluster.
+/// It handles discovery of the devices specified in a Configuration by calling `Discover` on
+/// all registered discovery handlers that are using the same protocol as specified in `Configuration.protocol.name.`
+/// For each device discovered by the discovery handlers, it creates a device plugin.
+/// If a device disappears, it deletes the associated instance after a grace period (for non-local devices).
+/// Note: Since this structure is automocked, the compiler does not seem to be able to confirm that all the
+/// methods are being used. Therefore, #[allow(dead_code)] has been added to all methods that are not invoked or
+/// tested on a DiscoveryOperator.
+#[derive(Clone)]
+pub struct DiscoveryOperator {
+    discovery_handler_map: RegisteredDiscoveryHandlerMap,
```

Comment:
some comments for each member of the struct would be helpful

Reply:
fixed in c7f3cb1. It prompted me to remove `device_plugin_path` from DiscoveryOperator as it was not needed to be passed to the `DevicePluginBuilder`, since it was made mockable.